src/wallet_manager.cpp

```diff
@@ -0,0 +1,556 @@
+
+    WalletPtr wallet = createWallet(context);
+
+    crypto::secret_key recovery_key;
+    std::string old_language = "English";
+
+    LOG_PRINT_L1("Restore seed is '" << seed << "'");
+
+    if (!crypto::ElectrumWords::words_to_bytes(seed, recovery_key, old_language))
+    {
+      LOG_PRINT_L1("Restore account failed due to electrum-style word list verification");
+      //TODO: error reporting
+      return;
+    }
+
+    crypto::secret_key secret_key = wallet->wallet.generateFromData(password, recovery_key, true);
+    const cryptonote::account_base& account = wallet->wallet.get_account();
+    std::string public_address = account.get_public_address_str(m_testnet);
+    std::string account_data = wallet->wallet.getAccountData(password);    
+    std::string view_key (&account.get_keys().m_view_secret_key.data[0], sizeof(account.get_keys().m_view_secret_key.data));
+    std::string seed;
+
+    if (!wallet->wallet.get_seed(seed))
+      throw std::runtime_error("Can't get seed for new wallet");
+
+    wallet->wallet.load_cache(getWalletCacheFileName(public_address));
+
+    std::string wallet_id = getContextWalletId(public_address);
+
+    registerWallet(context, wallet_id, wallet);    
+
+    rapidjson::Document json;
+    json.SetObject();
+    rapidjson::Value value(rapidjson::kStringType), result_value(rapidjson::kNumberType);
+
+    result_value.SetInt(0);
+    json.AddMember("Result", result_value, json.GetAllocator());
+
+    value.SetString(public_address.c_str(), public_address.length());
+    json.AddMember("Address", value, json.GetAllocator());
+
+    value.SetString(view_key.c_str(), view_key.length());
+    json.AddMember("ViewKey", value, json.GetAllocator());
+
+    value.SetString(account_data.c_str(), account_data.length());
+    json.AddMember("Account", value, json.GetAllocator());
+
+    value.SetString(seed.c_str(), seed.length());
+    json.AddMember("Seed", value, json.GetAllocator());
+
+    rapidjson::StringBuffer buffer;
+    rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+    json.Accept(writer);
+
+    result.body = buffer.GetString();
+
+    LOG_PRINT_L1("Wallet '" << public_address << "' has been successfully restored (callback=" << callback_url << ")");
+  });
+}
+
+void WalletManager::requestBalance(Context& context, const WalletId& wallet_id, const std::string& account_data, const std::string& password, const Url& callback_url)
+{
+  runAsyncForWallet(context, wallet_id, account_data, password, callback_url, [this, wallet_id, callback_url](tools::GraftWallet& wallet, OutHttp& result) {
+    LOG_PRINT_L1("Request balance for wallet '" << wallet_id << "'(callback=" << callback_url << ")");
+
+    rapidjson::Document json;
+    json.SetObject();
+    rapidjson::Value value(rapidjson::kNumberType);
+
+    value.SetInt(0);
+    json.AddMember("Result", value, json.GetAllocator());
+
+    value.SetDouble(wallet.balance() / double(ATOMS)); //todo: is conversion to double correct for currency?
```

Comment:
To answer to the todo question is: No.  The division here leads to rounding errors.  A wallet balance in any API should always been in lossless atomic units; a potentially lossy conversion from atomic to coin balance should only be done in the final display of the value to a user.

Reply:
I tend to agree with @jagerman

Reply:
Agree. However, WalletAPI requires double for prices. To avoid this I propose to change API and use integers for all balance fields.

Reply:
Fixed